v3.6 ignores "smart-tabs" rule and notifies of indent errors

I'm a maintainer of [Leaflet](https://github.com/Leaflet/Leaflet), and we use an [eslint config](https://github.com/Leaflet/Leaflet/blob/master/package.json#L44) with `smart-tabs`, like:

```
      "no-mixed-spaces-and-tabs": [2, "smart-tabs"],
      "indent": ["error", "tab", {"VariableDeclarator": 0}],
```

eslint 3.0 through 3.5 work just fine. However, after an upgrade to eslint 3.6, we've got error messages like the following:

```
/home/ivan/devel/Leaflet/src/control/Control.js
   64:7  error  Expected indentation of 2 tabs but found 4 spaces and 2 tabs  indent
```

This can be reproduced by checking out the [Leaflet repo](https://github.com/Leaflet/Leaflet), then running `npm install` and then `npm test`. Note that running `npm install eslint@3.5` then `npm test` will work as expected.

We don't use any transpiler in the Leaflet toolchain; full eslint config is available on the Leaflet repo; and the bug is reproducible on the environments of more than one developer.

Looking at eslint's changelog, I'll guess that this is caused by #7076 (ping @not-an-aardvark)

OK, I'm not used to your workflow so I wasn't fully aware that you expected a *minimal* example from me. Sorry about that.

I copy-pasted some of our [production code](https://github.com/Leaflet/Leaflet/blob/master/src/layer/Layer.js) and made this small reproducible example:

```
/* eslint indent:[2, "tab", {"VariableDeclarator": 0}], no-mixed-spaces-and-tabs:[2, "smart-tabs"] */

// Code snipped from https://github.com/Leaflet/Leaflet/blob/master/src/layer/Layer.js

L.Layer = L.Evented.extend({
	_updateZoomLevels: function () {
		var minZoom = Infinity,
		    maxZoom = -Infinity,
		    oldZoomSpan = this._getZoomSpan();

		for (var i in this._zoomBoundLayers) {
			var options = this._zoomBoundLayers[i].options;
			minZoom = options.minZoom === undefined ? minZoom : Math.min(minZoom, options.minZoom);
			maxZoom = options.maxZoom === undefined ? maxZoom : Math.max(maxZoom, options.maxZoom);
		}

		this._layersMaxZoom = maxZoom === -Infinity ? undefined : maxZoom;
		this._layersMinZoom = minZoom === Infinity ? undefined : minZoom;

		if (oldZoomSpan !== this._getZoomSpan()) {
			this.fire('zoomlevelschange');
		}
	}
});
```

It seems that tabs/spaces get lost in markdown formatting, so I've put [the code in a gist](https://gist.github.com/IvanSanchez/7529e0516ec260d4efe8ce510fab1083).

Note the spaces and tabs in the variables declaration:

```
→   →   var·minZoom·=·Infinity,
→   →   ····maxZoom·=·-Infinity,
→   →   ····oldZoomSpan·=·this._getZoomSpan();
```

Using eslint v3.5.0 (or any previous version), there are no linting errors. This is the expected behaviour.

Using eslint 3.6.1, I get the following output:

```
/tmp/eslint-smart-tabs-bug.js
   9:7  error  Expected indentation of 2 tabs but found 4 spaces and 2 tabs  indent
  10:7  error  Expected indentation of 2 tabs but found 4 spaces and 2 tabs  indent

✖ 2 problems (2 errors, 0 warnings)
```

--- a/lib/rules/indent.js
+++ b/lib/rules/indent.js
@@ -233,6 +233,12 @@ module.exports = {
          */
         function report(node, needed, gottenSpaces, gottenTabs, loc, isLastNodeCheck) {
 
+            if (gottenSpaces && gottenTabs) {
+
+                // To avoid conflicts with `no-mixed-spaces-and-tabs`, don't report lines that have both spaces and tabs.
+                return;
+            }
+
             const desiredIndent = (indentType === "space" ? " " : "\t").repeat(needed);
 
             const textRange = isLastNodeCheck

--- a/tests/lib/rules/indent.js
+++ b/tests/lib/rules/indent.js
@@ -1551,7 +1551,23 @@ ruleTester.run("indent", rule, {
             "      bar();\n" +
             "}",
             options: [2, {FunctionExpression: {parameters: "first", body: 3}}] // FIXME: make sure this is correct
-        }
+        },
+        {
+            code:
+            "function foo() {\n" +
+            "  bar();\n" +
+            "  \tbaz();\n" +
+            "\t   \t\t\t  \t\t\t  \t   \tqux();\n" +
+            "}",
+            options: [2]
+        },
+        {
+            code:
+            "function foo() {\n" +
+            "  bar();\n" +
+            "   \t\t}",
+            options: [2]
+        },
     ],
     invalid: [
         {
@@ -2840,35 +2856,19 @@ ruleTester.run("indent", rule, {
         {
             code:
             "var foo = bar;\n" +
-            "  \t  \t  \t  var baz = qux;",
+            "\t\t\tvar baz = qux;",
             output:
             "var foo = bar;\n" +
             "var baz = qux;",
             options: [2],
-            errors: expectedErrors([2, "0 spaces", "8 spaces and 3 tabs", "VariableDeclaration"])
-        },
-        {
-            code:
-            "function foo() {\n" +
-            "  bar();\n" +
-            "  \tbaz();\n" +
-            "\t   \t\t\t  \t\t\t  \t   \tqux();\n" +
-            "}",
-            output:
-            "function foo() {\n" +
-            "  bar();\n" +
-            "  baz();\n" +
-            "  qux();\n" +
-            "}",
-            options: [2],
-            errors: expectedErrors([[3, "2 spaces", "2 spaces and 1 tab", "ExpressionStatement"], [4, "2 spaces", "10 spaces and 9 tabs", "ExpressionStatement"]])
+            errors: expectedErrors([2, "0 spaces", "3 tabs", "VariableDeclaration"])
         },
         {
             code:
             "function foo() {\n" +
             "\tbar();\n" +
-            "\t  baz();\n" +
-            "  \t\t \t     \t   \t  \t\t\t qux();\n" +
+            "  baz();\n" +
+            "              qux();\n" +
             "}",
             output:
             "function foo() {\n" +
@@ -2877,19 +2877,7 @@ ruleTester.run("indent", rule, {
             "\tqux();\n" +
             "}",
             options: ["tab"],
-            errors: expectedErrors("tab", [[3, "1 tab", "2 spaces and 1 tab", "ExpressionStatement"], [4, "1 tab", "14 spaces and 8 tabs", "ExpressionStatement"]])
-        },
-        {
-            code:
-            "function foo() {\n" +
-            "  bar();\n" +
-            "\t\t}",
-            output:
-            "function foo() {\n" +
-            "  bar();\n" +
-            "}",
-            options: [2],
-            errors: expectedErrors([[3, "0 spaces", "2 tabs", "BlockStatement"]])
+            errors: expectedErrors("tab", [[3, "1 tab", "2 spaces", "ExpressionStatement"], [4, "1 tab", "14 spaces", "ExpressionStatement"]])
         }
     ]
 });

